[Integration] Fixes for integration tests

jsliacan opened the pull request:
**Fixes:** Issue #3771

**Related to** Issue #3796

Commit:
integration: run ginkgo nodes serially and in ordered fashion

Commit:
integration: refactor top-level testsuite info and actions

Commit:
integration: use default network mode for linux

cfergeau commented on test/integration/proxy_test.go:
```diff
@@ -35,7 +35,9 @@ var _ = Describe("", Serial, Ordered, Label("openshift-preset", "goproxy"), func
-			Expect(RunCRCExpectSuccess("config", "set", "network-mode", networkMode), ContainSubstring("Network mode"))
+			if runtime.GOOS == "linux" {
+				Expect(RunCRCExpectSuccess("config", "set", "network-mode", networkMode), ContainSubstring("Network mode"))
+			}
```
I'd argue that it's not very useful to test a non-default network mode, as this is not what most users will experience. Are there some issues if we run with the default network mode on linux?

gbraad replied:
@jsliacan WDYS?

jsliacan replied:
The reason why it was this way was mostly convenience - at the time when I wrote that test, you could set network mode on all platforms, so no extra code needed for linux. Also, linux system mode is subject to the e2e proxy test (which is only for linux). So this seemed to cover another (albeit less useful) case of user mode on linux.

That said, I see how it's a non-default option and thus might not be worth testing at this point. If you say that that's the case I'm happy to parametrize the proxy address by platform and use system mode on linux.

cfergeau replied:
>  If you say that that's the case I'm happy to parametrize the proxy address by platform and use system mode on linux.

Ah, if this makes the proxy test/code simpler, this can be a reason for using `user` mode networking on linux. Let's keep this for now to try to fix integration tests.

cfergeau replied:
With system networking, it's https://github.com/crc-org/crc/blob/85cc7183e4daf12b148b594a444e1d50546910e3/test/extended/util/proxy.go#L68 which would need to be changed to something reachable by the VM?

NB: should have looked at your recent changes before commenting ^^

cfergeau replied:
I'm surprised this works with user-mode networking though if you don't' set `* host-network-access                        Allow TCP/IP connections from the CRC VM to services running on the host (true/false, default: false)`

gbraad reviewed (commented):
overall OK. moving to a serial execution... Does this add to the total runtime?

the networkmode is an open question

Commit:
tests: move util functions to extended utils

Commit:
integration: wait for cluster in state

In e2e we use WaitForClusterInState after `crc start` finishes to
give time to the cluster to stabilize (all operators are reliably
up). This is also needed here before we login to the cluster.

cfergeau commented on test/integration/podman_test.go:
```diff
@@ -6,7 +6,7 @@ import (
 
 	"github.com/crc-org/crc/v2/test/extended/crc/cmd"
-	"github.com/crc-org/crc/test/extended/util"
+	"github.com/crc-org/crc/v2/test/extended/util"
```
This can be squashed in " integration: cleanup to AfterAll and BeforeSuite "

cfergeau reviewed (approved):
Really not clear why the tests would still be failing :( 
But looks good to me, ideally the last commit would be squashed to the right place.

Commit:
tests: avoid using shell struct in CheckCRCStatus

In integration tests we do not make use of ShellInstance, so in order
to be able to share WaitForClusterInState between e2e and integration,
we need CheckCRCStatus not to rely on ShellInstance. See also this
comment:
https://github.com/crc-org/crc/pull/3821#discussion_r1346340794.

jsliacan commented:
/retest

Commit:
integration: extend timeout

cfergeau commented on test/extended/crc/cmd/cmd.go:
```diff
@@ -140,11 +140,14 @@ func CheckCRCStatus(state string) error {
 
-	err := util.ExecuteCommand(CRC("status").ToString())
+	cmd := exec.Command("crc", "status")
+	out, err := cmd.Output()
```
Why do we want to avoid this?

jsliacan replied:
Because I need to explicitly get the stdout to check what was in it, and `util.ExecuteCommand` does not return that and isn't easily modifiable to do so (and even if it was, it's used in many many places which would need changing too). So going around it like above seemed better. I hope I am not overlooking something here, but that was the reasoning.

cfergeau replied:
looking at `util/shell.go` code, `CommandReturnShouldMatch("stdout"` calls `CompareExpectedWithActualContains(expected, shell.GetLastCmdOutput("stdout"))` which is supposed to get the command stdout output. Not clear why it's not working for you :-/

cfergeau replied:
I suggest keeping this one out of the PR so that we can merge it, and to address issues related to this in a follow-up PR.

jsliacan replied:
In integration tests, we do not use the shell construct (to simulate a user interacting with crc in one shell session). In e2e, we do (initialized [here](https://github.com/crc-org/crc/blob/02e1f697a0f8d8882e12684835a5ea1d683753fa/test/e2e/testsuite/testsuite.go#L157)): 
```go
err := util.StartHostShellInstance(testWithShell)
```
We had issues with running `crc setup` (with all the progress bars etc.) in e2e tests and had to move the execution of that command out of the general shell. So with integration, we didn't use the shell construct at all.

Given that we do not have an instance of [ShellInstance](https://github.com/crc-org/crc/blob/02e1f697a0f8d8882e12684835a5ea1d683753fa/test/extended/util/shell.go#L48) available, functions which rely on it won't work as expected. Therefore, instead of retrieving the stdout content from the `shell` variable that would contain it if it existed, we need to pass it between command execution (`crc status`) and output analysis (does it contain `Running`?) by hand. And that's what this commit does. Without it, `WaitForClusterInState` won't have any effect (except for wasting 10mins). 

So if we omit this change from this PR, integration tests will keep failing, albeit with a different error than initially when we started out.

cfergeau replied:
This would be a nice explanation to have in the commit log ;) 

`CheckCRCStatus` is used by both e2e and integration tests, is it ok to remove the use of `shell` for e2e?
And what about all the users of `ExecuteCommand()` in cmd.go? They also rely on `shell.go`, is this going to cause issues?

If this change is good enough to fix integration tests without breaking e2e, we can go with it, but some follow-up cleanups would be very useful imo (a clean separation between helpers using `exec.Command`, and helpers using `shell.go`). The former would only be used in integration tests, and the latter in e2e.

jsliacan replied:
> This would be a nice explanation to have in the commit log ;)

Will add it there.

> CheckCRCStatus is used by both e2e and integration tests, is it ok to remove the use of shell for e2e?

It is ok in the sense that it'll work. However, it'll dent the effort to simulate single-session user interaction with CRC.

> And what about all the users of ExecuteCommand() in cmd.go? They also rely on shell.go, is this going to cause issues?

As long as you don't need to retrieve anything aside from the error, it will do the job the same way.

> some follow-up cleanups would be very useful imo

I don't know the extent of this. It could have further consequences -- maybe it makes less sense to share the code between e2e and integration then. I opened an issue to track this: #3859

cfergeau replied:
> maybe it makes less sense to share the code between e2e and integration then

Maybe some go interface can be defined and used in the shared code. The interface would either make use of the shell code to run the command/get stdout/..., or would directly run the command for the integration tests.

cfergeau commented on test/extended/util/proxy.go:
```diff
@@ -68,7 +68,9 @@ func RunProxy() {
-	proxy.Logger = log.StandardLogger()
+	l := log.New()
+	l.Out = f
+	proxy.Logger = l
```
This seems out of place in `integration: cleanup to AfterAll and BeforeSuite`.
Is this required to fix something?

jsliacan replied:
Fair, it is out of place. I was reading something about logging and felt the need to change this asap :). Will remove it from this PR! Thanks

cfergeau reviewed (commented):
2 minor comments, otherwise looks good!

Commit:
integration: cleanup to AfterAll and BeforeSuite

The outer Describe block needs to be Ordered to be able to use
AfterAll inside it. Cleanup in BeforeSuite and AfterAll functions
should take care of clean environment before and after each testcase.

cfergeau commented on test/integration/podman_test.go:
```diff
@@ -6,11 +6,24 @@ import (
 )
 
-var _ = Describe("podman preset", Serial, Label("podman-preset"), func() {
+var _ = Describe("podman preset", Serial, Ordered, Label("podman-preset"), func() {
```
Is it intentional that the `Ordered` addition is in this `integration: cleanup to AfterAll and BeforeSuite` commit and not in `integration: run ginkgo nodes serially and in ordered fashion`? If it's not intentional, I've already split it in https://github.com/cfergeau/crc/commits/ginkgo-issue

jsliacan commented:
I forgot that I need to unset HTTPS_PROXY and HTTP_PROXY env variables that are implicitly set during the test.

cfergeau reviewed (approved):
I do not see https://github.com/crc-org/crc/pull/3821/commits/497c1d0000704e2822613402528ec721cde112fa#r1346340794 in the commit log for " tests: avoid using shell struct in CheckCRCStatus" but apart from this, ACK.

jsliacan commented:
Last changes include `git reset` onto @cfergeau 's branch with improved commits, additional `unset HTTP(S)_PROXY` in AfterAll, and updated commit message to explain why we're changing `CheckCRCStatus` to avoid using `ShellInstance`.

Commit:
integration: set CRCHome var before using it

Commit:
tests: remove typo

cfergeau commented on test/integration/podman_test.go:
```diff
@@ -50,7 +50,7 @@ var _ = Describe("podman preset", Serial, Ordered, Label("podman-preset"), func(
 				path = os.ExpandEnv(unexpandedPath)
 				csshk = filepath.Join(userHomeDir, ".crc/machines/crc/id_ecdsa")
-				dh = "npipe:////./pipe/rc-podman"
+				dh = "npipe:////./pipe/crc-podman"
```
This also happens in the E2E tests

jsliacan replied:
True, thanks. Grepped it now and the e2e occurrence was the last one. Changed it and updated the commit too.

cfergeau commented:
(waiting for integration-crc to run before merging it)

praveenkumar commented:
e2e-integration now succeed.
